main/lsp/LSPTypechecker.cc

```diff
@@ -670,22 +673,52 @@ std::vector<std::unique_ptr<core::Error>> LSPStaleTypechecker::retypecheck(std::
 
 LSPQueryResult LSPStaleTypechecker::query(const core::lsp::Query &q,
                                           const std::vector<core::FileRef> &filesForQuery) const {
-    ENFORCE(false, "query not implemented");
-    return LSPQueryResult();
+    // TODO: ENFORCE that we are on the indexing thread.
+    auto &gs = undoState.getEvictedGs();
+
+    // We assume gs is a copy of initialGS, which has had the inferencer & resolver run.
+    ENFORCE(gs->lspTypecheckCount > 0,
+            "Tried to run a query with a GlobalState object that never had inferencer and resolver runs.");
+
+    // Replace error queue with one that is owned by this thread.
+    auto queryCollector = make_shared<QueryCollector>();
+    gs->errorQueue = make_shared<core::ErrorQueue>(gs->errorQueue->logger, gs->errorQueue->tracer, queryCollector);
+
+    Timer timeit(config->logger, "query");
+    prodCategoryCounterInc("lsp.updates", "query");
+    ENFORCE(gs->errorQueue->isEmpty());
+    ENFORCE(gs->lspQuery.isEmpty());
+    gs->lspQuery = q;
+    auto resolved = getResolved(filesForQuery);
+    tryApplyDefLocSaver(*gs, resolved);
+    tryApplyLocalVarSaver(*gs, resolved);
+
+    pipeline::typecheck(gs, move(resolved), config->opts, *emptyWorkers, /*presorted*/ true);
```

Comment:
This is not safe, right?  We don't want `undoState.getEvictedGs` to be modified, yet we're passing it to a function that's going to modify it.  (`getResolved` has the same issue.)

Reply:
Since we're pulling in the evicted file, which was already indexed relative to the evicted gs, I think we shouldn't actually be introducing any new symbols and therefore this should be de facto safe. (Not 100% confident though, so I should also note that this is hidden behind an opt-in CLI flag.)

Reply:
I don't know that we have an issue with `typecheck`, because `typecheck` already works with a const GlobalState. I think that the only problem is with `getResolved`.

Reply:
I think that the idea of using `evictedGs` is a clever idea to not need to borrow the up-to-date GlobalState from the typechecker thread, but I think this confirms that we still want to gate landing this on the non-mutating namer and resolver.

Reply:
> I think this confirms that we still want to gate landing this on the non-mutating namer and resolver.

Yeah, I'd defer to your judgment on that but I think it's probably a fair call to make.

Reply:
The types here are a bit gross, because `typecheck` doesn't really modify the global state (except for bits that munge the `errorQueue`, but that's already marked `mutable` anyway, but `typecheck` is not typed as such.  I'll fix that up someplace else.